[source-postgres/mssql/mysql] Send state and count for full refresh

xiaohansong opened the pull request:
## What
Non RFR stream on postgres will now send state message with count.

Will only send one final state message for these streams.

## How
<!--
* Describe how code changes achieve the solution.
-->

## Review guide
<!--
1. `x.py`
2. `y.py`
-->

## User Impact
<!--
* What is the end result perceived by the user?
* If there are negative side effects, please list them. 
-->

## Can this PR be safely reverted and rolled back?
<!--
* If unsure, leave it blank.
-->
- [ ] YES 💚
- [ ] NO ❌

Commit:
postgres send state for full refresh

Commit:
Merge remote-tracking branch 'origin/master' into xiaohan/statefullrefresh

Commit:
java

Commit:
local cdk

theyueli commented on airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/source/relationaldb/state/SourceStateIterator.kt:
```diff
@@ -84,7 +87,7 @@ open class SourceStateIterator<T>(
      */
     private fun shouldAttachCountWithState(): Boolean {
-        return stream?.syncMode != SyncMode.FULL_REFRESH
+        return true
```
does this mean we no longer need this function?

akashkulk replied:
+1 this should be removed here and below

theyueli commented on airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/source/relationaldb/state/SourceStateIterator.kt:
```diff
@@ -71,6 +73,7 @@ open class SourceStateIterator<T>(
                 )
             }
             recordCount = 0L
+            LOGGER.info { "sending real state message: $finalStateMessageForStream" }
```
nit: what does "real" mean here :-)

xiaohansong replied:
debug log - will remove! thanks for catching it.

theyueli commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/CtidCountStateManager.java:
```diff
@@ -0,0 +1,81 @@
+
+/**
+ * A state manager for non resumable full refresh. It does not produce any meaningful state, but
+ * just serves as feeding into SourceStateIterator so we can have counts.
```
serves as "input" ?

theyueli commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/CtidGlobalStateManager.java:
```diff
@@ -50,25 +51,34 @@ public CtidGlobalStateManager(final CtidStreams ctidStreams,
-            new AirbyteStreamNameNamespacePair(configuredAirbyteStream.getStream().getName(), configuredAirbyteStream.getStream().getNamespace()));
+        this.resumableFullRefreshStreams.add(pair);
+      } else if (configuredAirbyteStream.getSyncMode() == SyncMode.FULL_REFRESH) {
+        this.nonResumableFullRefreshStreams.add(pair);
```
This can be refactored to be the following:
```
if (configuredAirbyteStream.getSyncMode() == SyncMode.FULL_REFRESH) {
  if (fileNodeHandler.hasFileNode(
          new AirbyteStreamNameNamespacePair(configuredAirbyteStream.getStream().getName(), configuredAirbyteStream.getStream().getNamespace()))
    this.resumableFullRefreshStreams.add(pair);
  else
    this.nonResumableFullRefreshStreams.add(pair);
}
```

theyueli commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/CtidGlobalStateManager.java:
```diff
@@ -50,25 +51,34 @@ public CtidGlobalStateManager(final CtidStreams ctidStreams,
       }
     });
+
+    LOGGER.info("*** resumable full refresh: " + resumableFullRefreshStreams);
```
nit: why do we need these fancy stars "***"

xiaohansong replied:
oops - fancy stars are just used for debug logs, so i can easily search for it. will remove!

theyueli commented on airbyte-integrations/connectors/source-postgres/src/test/java/io/airbyte/integrations/source/postgres/CdcPostgresSourceTest.java:
```diff
@@ -84,11 +85,9 @@ protected void setBaseImage() {
-    // stateMessages.stream().forEach(stateMessage ->
-    // count.addAndGet(stateMessage.getSourceStats().getRecordCount().longValue()));
-    // assertEquals(totalCount, count.get());
+    AtomicLong count = new AtomicLong(0L);
```
Just a question for my curiosity: since we don't have any concurrent operations in our implementation, why do we need an atomic variable?

xiaohansong replied:
likely not needed at all - as long as we can compare the count I think we should be good

xiaohansong replied:
oh i remembered - lambda function doesn't like me defining a long count = 0 then count += stateMessage.getCount() because it needs to be final

Commit:
Merge remote-tracking branch 'origin/master' into xiaohan/statefullrefresh

btkcodedev committed:
🐛Source Square: Add error handlers, migrate to inline schema and move spec (#39377)

anjay-goel committed:
🐛 Source Appsflyer: Fix organic In-app events stream (#39407)

Co-authored-by: marcosmarxm <marcosmarxm@gmail.com>

Commit:
🐛 Source Stripe: fix the `invoices` stream, incremental sync, after the PR #39393 (#39412)

Commit:
clean up

akashkulk commented on airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/kotlin/io/airbyte/cdk/integrations/debezium/CdcSourceTest.kt:
```diff
@@ -781,7 +781,10 @@ abstract class CdcSourceTest<S : Source, T : TestDatabase<*, T, *>> {
             )
         } else {
-            assertExpectedStateMessageCountMatches(stateMessages1, MODEL_RECORDS.size.toLong())
+            assertExpectedStateMessageCountMatches(
```
nit : Can you add a quick comment as to why (since MODEL_RECORDS_2) is sending state now

nataliekwong committed:
docs: Add icons to Connection/Stream Status (#39409)

akashkulk commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/CtidCountStateManager.java:
```diff
@@ -0,0 +1,81 @@
+ * A state manager for non resumable full refresh. It does not produce any meaningful state, but
+ * just serves as feeding into SourceStateIterator so we can have counts.
+ */
+public class CtidCountStateManager implements SourceStateMessageProducer<AirbyteMessage> {
```
Can we have "non-resumable" in the name somewhere?

xiaohansong replied:
Sure. While doing this change I actually realized I can refactor this into CDK to make things much cleaner!

nataliekwong committed:
docs: May 2024 Release Notes (#39385)

Co-authored-by: Audrey Maldonado <audrey.maldonado@gmail.com>

akashkulk commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/ctid/CtidGlobalStateManager.java:
```diff
@@ -50,23 +51,29 @@ public CtidGlobalStateManager(final CtidStreams ctidStreams,
+            new AirbyteStreamNameNamespacePair(configuredAirbyteStream.getStream().getName(), configuredAirbyteStream.getStream().getNamespace()))) {
+          this.resumableFullRefreshStreams.add(pair);
+        } else {
+          this.nonResumableFullRefreshStreams.add(pair);
```
Can we add logs/analytics messages for number of non resumable full refresh streams? I'd be curious how often they are encountered

xiaohansong replied:
Sure - I'm curious how do we do that?
Alternatively I guess we can also query platform database (config-db) to see how many streams has that is_resumable flag on catalog

akashkulk replied:
Here is an example of how to emit it: https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialReadUtil.java#L153

You need to define an analytics message in this file : https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/db/DbAnalyticsUtils.kt

Once that's in, I can show you how to create a dashboard from this!

akashkulk replied:
Not blocking - I'll leave it up to you on what you think works since as you pointed out, this information is also available. by querying platform DB

xiaohansong replied:
Got it, thanks for the pointer! 

I think maybe querying db fits this more. We can know who run them and also how often they are running

akashkulk reviewed (commented):
Minor nits, looks good otherwise

gosusnp committed:
Bump Airbyte version from 0.62.4 to 0.63.0

alafanechere committed:
source-salesforce: [autopull] Upgrade base image to v1.2.2 (#39269)

Co-authored-by: Baz <oleksandr.bazarnov@globallogic.com>

alafanechere committed:
source-zendesk-sell: [autopull] Upgrade base image to v1.2.2 (#39258)

btkcodedev committed:
✨Source Secoda: Make Connector Compatible with Builder (#38932)

abedzantout committed:
[source-xero] Add OAuth and Bearer Strategies (#39316)

Co-authored-by: marcosmarxm <marcosmarxm@gmail.com>

tolik0 committed:
chore(source-azure-blob-storage): fix abnormal state for `parquet` stream (#39427)

pabloescoder committed:
✨Source Salesloft: migrate to low-code (#38362)

Co-authored-by: Alexandre Girard <alexandre@airbyte.io>

dependabot[bot] committed:
build(deps-dev): bump braces from 3.0.2 to 3.0.3 in /airbyte-integrations/connector-templates/generator (#39421)

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

lazebnyi committed:
Source Partnerstack: Chenged last records to last record (#37595)

Co-authored-by: Marcos Marx <marcosmarxm@users.noreply.github.com>

bnchrch committed:
feat(docs): Add cdk version, metrics, and last updated (#38799)

## What

Adds the following to the metrics component in docs
1. CDK version
2. Last updated ad
3. Sync success rate
4. Usage rate


Does not
1. Add metric icons. (Next PR

juftin committed:
🐛 Source Square: fix customers pagination (#30315)

Co-authored-by: Mal Hancock <mallory@archangelic.space>
Co-authored-by: Marcos Marx <marcosmarxm@users.noreply.github.com>
Co-authored-by: marcosmarxm <marcosmarxm@gmail.com>

bnchrch committed:
feat(docs): Add icon component for metrics (#39104)

## What
Adds the overlapping icons from the figma mockup
https://www.figma.com/design/JX15SrsqQK53Gm3Ekg1Dq5/Connector-Quality-Indicators?node-id=78-144&t=Ea6R18ejiYHfEEZi-1

kikoncuo committed:
🐛 Postgres Destination fails when the connection is persisted across steps (#39372)

bnchrch committed:
bug(docs): Update mobile style of CQM component (#39435)

gndaniela committed:
🐛 Source Klarna: fixing connector compatibility with the US API endpoint (#38709)

Co-authored-by: Marcos Marx <marcosmarxm@users.noreply.github.com>
Co-authored-by: marcosmarxm <marcosmarxm@gmail.com>

carlonuccio committed:
🐛 Source Kafka: fix empty airbyte_data after airbyte version 0.50.4 for AvroFormat (#32538)

Co-authored-by: Marcos Marx <marcosmarxm@users.noreply.github.com>
Co-authored-by: marcosmarxm <marcosmarxm@gmail.com>

lazebnyi committed:
Source Datadog: Chaged last records to last record (#37590)

bnchrch committed:
feat(docs): update metric icon (#39449)

Commit:
apply changes to mssql and mysql

Commit:
remove unwanted lock

pabloescoder committed:
✨Source Linkedin Pages: migrate to low-code (#36744)

Co-authored-by: Natik Gadzhi <natik@respawn.io>

ctdio committed:
fix(source_slack): use private channels config when reading slack threads (#39416)

Co-authored-by: Natik Gadzhi <natik@respawn.io>

Commit:
fix cdc state

theyueli committed:
source-mysql: upgrade debezium to 2.5.4 (#39144)

This patch attempts to fix the following issues:

airbytehq/oncall#5332
#32348

cmm-airbyte committed:
Increasing maxSecondsBetweenMessages for facebook marketing connector o avoid heartbeat timeout error (#39400)

Co-authored-by: cristina.mariscal <cristina.mariscal@cristina.mariscal--MacBook-Pro---DFJ27FJFXX>

Commit:
fix bugs and test

pabloescoder committed:
✨Source Babelforce: Make Connector Compatible with Builder (#38776)

pabloescoder committed:
✨Source Workramp: Make Connector Compatible with Builder (#38741)

pabloescoder committed:
✨Source News Api: Make Connector Compatible with Builder (#38635)

gvillafanetapia committed:
✨ Source Google Analytics 4: make lookback window configurable (#38884)

theyueli committed:
CDK: translate java EOF exception's message to a more customer-friendly message (#39417)

fixes airbytehq/oncall#4283

jdenquin committed:
✨ [source-getLago] Add Fees Support (#35661)

Co-authored-by: marcosmarxm <marcosmarxm@gmail.com>

alafanechere committed:
airbyte-ci: make slugify consistent with live test (#39457)

maxi297 committed:
🐛 Source Facebook Marketing: apply configured statuses to ads insight streams (#39439)

Co-authored-by: williammcguinness <wmcguinness@findoctave.com>

Commit:
source-postgres: connect with adaptiveFetch=true, take 2 (#38773)

stephane-airbyte committed:
destination-azure: fix AzureBlobStorageSpillTest (#39465)

test was passing when run from IDE, but not when run with airbyte-ci. Issue was the hardcoding of the container's IP

gisripa committed:
Destination Postgres: Kotlin auto conversion src/main/java -> src/main/kotlin (#39388)

tolik0 committed:
chore(source-sendgrid): fix type of cursor in abnormal state (#39469)

akashkulk committed:
[DB sources] : Debezium properties accept list of streams (#39405)

btkcodedev committed:
✨Source Senseforce: Make Connector Compatible with Builder (#39462)

btkcodedev committed:
✨Source RocketChat: Make Connector Compatible with Builder (#39110)

natikgadzhi committed:
source-youtube-analytics: [autopull] base image + poetry + up_to_date (#38546)

Co-authored-by: Danylo Jablonski <150933663+DanyloGL@users.noreply.github.com>
Co-authored-by: Augustin <augustin@airbyte.io>
Co-authored-by: Serhii Lazebnyi <53845333+lazebnyi@users.noreply.github.com>
Co-authored-by: Serhii Lazebnyi <serhii.lazebnyi@globallogic.com>
Co-authored-by: alafanechere <augustin.lafanechere@gmail.com>

natikgadzhi committed:
source-exchange-rates: [autopull] base image + poetry + up_to_date (#38543)

Co-authored-by: Serhii Lazebnyi <serhii.lazebnyi@globallogic.com>
Co-authored-by: Serhii Lazebnyi <53845333+lazebnyi@users.noreply.github.com>
Co-authored-by: alafanechere <augustin.lafanechere@gmail.com>
Co-authored-by: Danylo Jablonski <150933663+DanyloGL@users.noreply.github.com>
Co-authored-by: Augustin <augustin@airbyte.io>

natikgadzhi committed:
source-paystack: [autopull] base image + poetry + up_to_date (#38430)

Co-authored-by: Danylo Jablonski <150933663+DanyloGL@users.noreply.github.com>
Co-authored-by: Augustin <augustin@airbyte.io>

theyueli commented on airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/integrations/source/relationaldb/state/NonResumableStateMessageProducer.kt:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    /**
+     * @param stream
```
Unfinished comments.

theyueli commented:
nit: need to update PR title since now the PR changes many connectors.

Commit:
pr fix

Commit:
Merge remote-tracking branch 'origin/master' into xiaohan/statefullrefresh

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/airbytehq/airbyte?pullRequest=39349) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/airbytehq/airbyte?pullRequest=39349) before we can accept your contribution.<br/>**26** out of **27** committers have signed the CLA.<br/><br/>:white_check_mark: btkcodedev<br/>:white_check_mark: xiaohansong<br/>:white_check_mark: nataliekwong<br/>:white_check_mark: bazarnov<br/>:white_check_mark: abedzantout<br/>:white_check_mark: alafanechere<br/>:white_check_mark: gosusnp<br/>:white_check_mark: lazebnyi<br/>:white_check_mark: bnchrch<br/>:white_check_mark: pabloescoder<br/>:white_check_mark: tolik0<br/>:white_check_mark: AnjayGoel<br/>:white_check_mark: juftin<br/>:white_check_mark: gndaniela<br/>:white_check_mark: carlonuccio<br/>:white_check_mark: kikoncuo<br/>:white_check_mark: gvillafanetapia<br/>:white_check_mark: theyueli<br/>:white_check_mark: jdenquin<br/>:white_check_mark: postamar<br/>:white_check_mark: ctdio<br/>:white_check_mark: maxi297<br/>:white_check_mark: gisripa<br/>:white_check_mark: natikgadzhi<br/>:white_check_mark: akashkulk<br/>:white_check_mark: stephane-airbyte<br/>:x: cmm-airbyte<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/airbytehq/airbyte?pullRequest=39349) it.</sub>

Commit:
bump version

Commit:
fix a test

Commit:
fix a merge error

Commit:
format

Commit:
Merge remote-tracking branch 'origin/master' into xiaohan/statefullrefresh

Commit:
fix an error

Commit:
Merge remote-tracking branch 'origin/master' into xiaohan/statefullrefresh

xiaohansong commented:
/publish-java-cdk

Commit:
Merge branch 'master' into xiaohan/statefullrefresh

Commit:
upgrade cdk dependency
